### DISCUSSION TITLE
Custom parts javascript?

### DISCUSSION THREAD

**Discussion by tlbrack** [SpinaCMS/Spina#1101] [Q&A] [answered]
When writing custom parts, what's the way to get javascript files (such as stimulus controllers) included properly in the import map used by the Spina Engine?

**Comment by Bramjetten** [comment, marked as answer]
This is how we do that in Spina Pro:

```ruby
# engine.rb
module Spina
  module Pro
    class Engine < ::Rails::Engine
      # ...
      initializer "spina.pro.assets.assets" do |app|
        # Add pro Stimulus controllers to Spina's importmap
        Spina.config.importmap.draw do |paths|
          pin_all_from Spina::Pro::Engine.root.join("app/assets/javascripts/spina/pro/controllers"), under: "controllers", to: "spina/pro/controllers"
          pin_all_from Spina::Pro::Engine.root.join("app/assets/javascripts/spina/pro/libraries"), under: "libraries", to: "spina/pro/libraries"
        end
        
        # Add assets manifest to assets.precompile
        app.config.assets.precompile += %w(spina/pro/manifest.js)
      end
    end
  end
end
```

**Reply by tlbrack** [reply]
Thanks!

I was attempting something similar in my app, but I think ran into issues because of the location of my custom controllers.  They are under `app/javascript/spina/controllers` rather than the `app/assets/...` path and were therefore not being included in the pre-compilation.

After adding another custom manifest file with a link_tree for them and appending that manifest to the precompile list, it seems to be working!
